readium/streamer/src/main/java/org/readium/r2/streamer/ParserAssetFactory.kt

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.streamer
+
+import java.io.File
+import java.nio.charset.Charset
+import org.json.JSONObject
+import org.readium.r2.shared.asset.Asset
+import org.readium.r2.shared.error.*
+import org.readium.r2.shared.fetcher.ContainerFetcher
+import org.readium.r2.shared.fetcher.ResourceFetcher
+import org.readium.r2.shared.fetcher.RoutingFetcher
+import org.readium.r2.shared.publication.Link
+import org.readium.r2.shared.publication.Manifest
+import org.readium.r2.shared.publication.Publication
+import org.readium.r2.shared.resource.Container
+import org.readium.r2.shared.resource.Resource
+import org.readium.r2.shared.util.http.HttpClient
+import org.readium.r2.shared.util.http.HttpFetcher
+import org.readium.r2.shared.util.mediatype.MediaType
+import org.readium.r2.shared.util.mediatype.MediaTypeRetriever
+import org.readium.r2.streamer.parser.PublicationParser
+
+internal class ParserAssetFactory(
+    private val httpClient: HttpClient,
+    private val mediaTypeRetriever: MediaTypeRetriever,
+) {
+
+    suspend fun createParserAsset(
+        asset: Asset,
+    ): Try<PublicationParser.Asset, Publication.OpeningException> {
+        return when (asset) {
+            is Asset.Container ->
+                Try.success(createFetcherForContainer(asset.container, asset.mediaType, asset.name))
+            is Asset.Resource ->
+                createFetcherForResource(asset.resource, asset.mediaType, asset.name)
+        }
+    }
+
+    private fun createFetcherForContainer(
+        container: Container,
+        mediaType: MediaType,
+        assetName: String
+    ): PublicationParser.Asset {
+        val fetcher = ContainerFetcher(container, mediaTypeRetriever)
+        return PublicationParser.Asset(assetName, mediaType, fetcher)
+    }
+
+    private suspend fun createFetcherForResource(
+        resource: Resource,
+        mediaType: MediaType,
+        assetName: String
+    ): Try<PublicationParser.Asset, Publication.OpeningException> =
+        if (mediaType.isRwpm) {
+            createFetcherForManifest(resource, assetName)
+        } else {
+            createFetcherForContent(resource, mediaType, assetName)
+        }
+
+    private suspend fun createFetcherForManifest(
+        resource: Resource,
+        assetName: String
+    ): Try<PublicationParser.Asset, Publication.OpeningException> {
+        val manifest = resource.readAsRwpm(packaged = false)
```

Comment:
What was the rationale for converting a manifest to a package asset here instead of in the `ReadiumWebPubParser` directly?

I'm concerned that this is not very extensible (e.g. if there's a format similar to RWPM that needs to be supported, W3C Web Publications for example). It also feels a bit weird to have RWPM specific code here.

Reply:
In `ReadiumWebPubParser` it was too late to create a `Fetcher` because `ContentProtection` might have modified it before. It's no longer the case: using manifests seems incompatible with `ContentProtection`s unless a `ContentProtection` itself implements this stuff.

I feel like this might deserve more thought though, let's postpone this change.

Reply:
OK, I'll add an entry in the V3 GH project